File: python/taflow/minus_directional_indicator.py
from typing import Any
import numpy as np
from ._native import MinusDirectionalIndicator as _Native
from ._series import as_float64_series
class MinusDirectionalIndicator:
    """Stateful MinusDirectionalIndicator indicator.
    Parameters are documented by the constructor signature; scalar
    ``append`` returns the current value and ``compute`` returns
    the aligned history with NaN warm-up where applicable.
    """
    def __init__(self, high:Any|None=None, low:Any|None=None, close:Any|None=None, timeperiod:int=14):
        self._state=_Native(timeperiod)
        if high is not None or low is not None or close is not None:self.extend(high,low,close)
    def append(self,h:float,l:float,c:float):self._state.append(h,l,c);return self
    def extend(self,h:Any,l:Any|None=None,c:Any|None=None):
        if l is None or c is None:raise ValueError("high, low, and close must be provided together")
        self._state.extend(as_float64_series(h),as_float64_series(l),as_float64_series(c));return self
    def compute(self)->np.ndarray:return self._state.compute()
    @property
    def value(self):return self._state.value
    def reset(self):self._state.reset();return self

File: crates/taflow-core/src/stream/minus_directional_indicator.rs
//! Incremental Minus Directional Indicator (-DI).
use super::directional::DirectionalMovement;
use crate::error::TaResult;

/// Persistent Rust state or aligned output type for `MinusDirectionalIndicator`.
///
/// The state consumes chronological inputs causally, preserves warm-up
/// values, and exposes the current result through its public API.
pub struct MinusDirectionalIndicator {
    directional: DirectionalMovement,
    value: Option<f64>,
}
impl MinusDirectionalIndicator {
    /// Computes or updates `new` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn new(period: usize) -> TaResult<Self> {
        Ok(Self {
            directional: DirectionalMovement::new(period)?,
            value: None,
        })
    }
    /// Computes or updates `append` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn append(&mut self, high: f64, low: f64, close: f64) -> Option<f64> {
        self.value = self
            .directional
            .append(high, low, close)
            .map(|v| v.minus_di);
        self.value
    }

    /// Append aligned HLC slices while preserving scalar state and warm-up.
    pub fn extend_slices_into(
        &mut self,
        high: &[f64],
        low: &[f64],
        close: &[f64],
        output: &mut Vec<f64>,
    ) -> TaResult<()> {
        if high.len() != low.len() || high.len() != close.len() {
            return Err(crate::TaError::LengthMismatch {
                expected: high.len(),
                got: low.len().min(close.len()),
            });
        }
        output.reserve(high.len());
        for ((high, low), close) in high.iter().zip(low).zip(close) {
            output.push(self.append(*high, *low, *close).unwrap_or(f64::NAN));
        }
        Ok(())
    }
    /// Computes or updates `value` through the native Rust kernel.
    ///
    /// Parameters are the typed series and configuration values in the signature.
    ///
    /// Returns the computed value, aligned history, or a validation error.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Reset the persistent state and clear the latest value.
    pub fn reset(&mut self) {
        self.directional.reset();
        self.value = None;
    }
}
